src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/RazorRenameFeature.ts

```diff
@@ -0,0 +1,96 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+import * as vscode from 'vscode';
+import { getRazorDocumentUri, isRazorCSharpFile } from './RazorConventions';
+import { RazorDocumentManager } from './RazorDocumentManager';
+import { RazorDocumentSynchronizer } from './RazorDocumentSynchronizer';
+import { RazorLanguage } from './RazorLanguage';
+import { RazorLanguageServiceClient } from './RazorLanguageServiceClient';
+import { RazorLogger } from './RazorLogger';
+import { RazorRenameProvider } from './RazorRenameProvider';
+import { LanguageKind } from './RPC/LanguageKind';
+
+export class RazorRenameFeature {
+    public readonly serviceClient: RazorLanguageServiceClient;
+    public readonly renameProvider: RazorRenameProvider;
+
+    constructor(
+        documentSynchronizer: RazorDocumentSynchronizer,
+        documentManager: RazorDocumentManager,
+        serviceClient: RazorLanguageServiceClient,
+        logger: RazorLogger) {
+            this.serviceClient = serviceClient;
+            this.renameProvider = new RazorRenameProvider(
+                documentSynchronizer,
+                documentManager,
+                serviceClient,
+                logger);
+    }
+
+    public register() {
+        const registrations = [
+            vscode.languages.registerRenameProvider(RazorLanguage.id, this.renameProvider),
+            vscode.commands.registerCommand(
```

Comment:
I forget how VSCode handles exceptions thrown from registered commands. When we make the changes on the O# side we'll probably need to bake in exception handling to ensure nothing blows up if exceptions propagate.

Reply:
They already handle exceptions. The code we write in OmniSharp will be inside a try block.